### ISSUE TITLE
ConversationalRetrievalQAChain factory method only supports QA stuff chain, fails on long inputs

### ISSUE THREAD

**Issue by Iambihua** [hwchase17/langchainjs#1417]
![image](https://github.com/hwchase17/langchainjs/assets/49816387/bd038e08-e298-4cce-8668-daa1213244a8)
**Here is a snippet of my code**:
```
const loader = new PDFLoader('./基于贝叶斯优化的无标签网络剪枝算法.pdf')
const pdf = await loader.load()
const textSplitter = new RecursiveCharacterTextSplitter({
  chunkSize: 1000,
  chunkOverlap: 0
})
const docs = await textSplitter.splitDocuments(pdf)
const vectorStore = await PineconeStore.fromDocuments(
  docs,
  new OpenAIEmbeddings({
    openAIApiKey: process.env.OPEN_API_KEY
  }),
  {
    pineconeIndex
  }
)`
`const vectorStore = await PineconeStore.fromExistingIndex(
  new OpenAIEmbeddings({
    openAIApiKey: process.env.OPEN_API_KEY
  }),
  { pineconeIndex }
)
/* Create the chain */
const chain = ConversationalRetrievalQAChain.fromLLM(
  model,
  vectorStore.asRetriever(),
  { verbose: true }
)
const question =
  '对于没有经过任何优化的高斯过程，n 个样本点时间复杂度大概是多少?'
try {
  // your async function here
  const res = await chain.call({ question, chat_history: [] })
  console.log(res.text)
} catch (error) {
  if (error.response) {
    console.log(error.response.status)
    console.log(error.response.data)
  } else {
    console.log(error.data.message)
  }
}
```

**Comment by Iambihua** [comment]
After adding the parameter verbose:true, I received the following error message in the terminal:
![image](https://github.com/hwchase17/langchainjs/assets/49816387/f70d78e0-5e7e-4ae1-b4b4-5abe2e0a2ea2)

**Comment by jacoblee93** [comment]
The error message you're seeing means that your prompt is too many tokens, and it looks like that's happening because the `ConversationalRetrievalQAChain.fromLLM` method is only letting you use a specific type of chain with no option to change it:

https://github.com/hwchase17/langchainjs/blob/main/langchain/src/chains/conversational_retrieval_chain.ts#L156

For now, can you try creating the conversation chain similarly to this?

```typescript
import { loadQAMapReduceChain } from "langchain/chains/load";
const question_generator_template = `Given the following conversation and a follow up question, rephrase the follow up question to be a standalone question.

Chat History:
{chat_history}
Follow Up Input: {question}
Standalone question:`;

const qa_template = `Use the following pieces of context to answer the question at the end. If you don't know the answer, just say that you don't know, don't try to make up an answer.

{context}

Question: {question}
Helpful Answer:`;

const  chain = new ConversationalRetrievalQAChain({
      retriever: vectorStore.asRetriever(),
      combineDocumentsChain: loadQAMapReduceChain(model),
      questionGeneratorChain: new LLMChain({
        prompt: question_generator_prompt,
        llm: model,
        verbose: true,
      }),
      verbose: true,
    });
```

Basically, using the constructor manually with a map reduce chain instead of the default stuff chain: https://github.com/hwchase17/langchainjs/blob/main/langchain/src/chains/conversational_retrieval_chain.ts#L62

**Comment by Iambihua** [comment]
Thank you for your answer. The solution you provided is very helpful!

**Comment by jacoblee93** [comment]
Awesome! I'm going to leave this open though because this should be configurable.

**Comment by Iambihua** [comment]
Indeed! However, I have another small question that I would like to ask you for advice on. My idea is, if the answer cannot be found from the PDF materials I provided, how can I make it turn to call an API to generate answers from OpenAI and also inform the user that the answer is not from the PDF database? I tried adjusting prompts but it never meets my expectations.

**Comment by jacoblee93** [comment]
In that case,
```
qaChainOptions: {
  type: "refine",
  template: YOUR_CUSTOM_TEMPLATE
}
```
The template will need to have certain input variables filled in though.

**Comment by Iambihua** [comment]
**_我修改了template和question_generator_prompt，但是它并没有起效_**

![image](https://github.com/hwchase17/langchainjs/assets/49816387/0365e57c-816c-494d-bb2b-2a0e179e4cfe)

**Comment by Iambihua** [comment]
**I modified the template and question_generator_prompt, but it didn't take effect.**

![image](https://github.com/hwchase17/langchainjs/assets/49816387/b543f144-dc39-4e93-ae22-91b3c3eb2971)

**Comment by jacoblee93** [comment]
Ah, my mistake. You'll need to pass `prompt` into `qaChainOptions`:

```
{
  type: "refine",
  prompt: PromptTemplate.fromTemplate(qa_template)
}
```

You should remove the other `qaTemplate`. Are you using Typescript?

**Comment by jacoblee93** [comment]
This is what it does internally: 
https://github.com/hwchase17/langchainjs/blob/90f6cb05e380dc45b1ae3f4e0d4287f19da0b2ef/langchain/src/chains/conversational_retrieval_chain.ts#L178

**Comment by Iambihua** [comment]
I used JavaScript and modified the code according to the method you provided, but there was not much difference compared to last time.

**Comment by jacoblee93** [comment]
Sorry, my mistake again. This is the object it takes:

https://github.com/hwchase17/langchainjs/blob/main/langchain/src/chains/question_answering/load.ts#L96

Since the refine chain works in two steps, there are two prompt options you can override. You can see the default values here: https://github.com/hwchase17/langchainjs/blob/main/langchain/src/chains/question_answering/refine_prompts.ts

```
{
  type: "refine",
  questionPrompt: PromptTemplate.fromTemplate(CUSTOM_QUESTION_TEMPLATE),
  refinePrompt: PromptTemplate.fromTemplate(CUSTOM_REFINE_TEMPLATE)
}
```

**Comment by Iambihua** [comment]
It's okay, I've got it covered. I really appreciate your help!

**Comment by Iambihua** [comment]
I have made multiple modifications to the prompt, but it did not return the expected results. Did I select the corresponding prompt incorrectly? Which part of the chain should I modify for the prompt? I hope that if it cannot find an answer in the knowledge base, it can directly call on a large model to provide an answer instead of relying on information from the knowledge base.
